integration/mediation-tests/tests-common/integration-test-utils/src/main/java/org/wso2/esb/integration/common/utils/servers/GreenMailServer.java

```diff
@@ -0,0 +1,244 @@
+     * Connecting to the imap mail store.
+     *
+     * @return imap store connection
+     * @throws ESBMailTransportIntegrationTestException
+     */
+    private static Store getImapConnection() throws ESBMailTransportIntegrationTestException, MessagingException {
+        Properties props = new Properties();
+        Session session = Session.getInstance(props);
+        URLName urlName = new URLName(PROTOCOL_IMAP, BIND_ADDRESS, 3143, null, greenMailUser.getLogin(),
+                greenMailUser.getPassword());
+        Store store = session.getStore(urlName);
+        store.connect();
+        return store;
+    }
+
+    /**
+     * Check mail folder for an email using subject.
+     *
+     * @param emailSubject
+     * @param folder
+     * @param protocol
+     * @return
+     * @throws ESBMailTransportIntegrationTestException
+     */
+    private static boolean isMailReceivedBySubject(String emailSubject, String folder, String protocol)
+            throws ESBMailTransportIntegrationTestException, MessagingException {
+        boolean emailReceived = false;
+        Folder mailFolder;
+        Store store;
+        if (PROTOCOL_POP3.equals(protocol)) {
+            store = getPopConnection();
+        } else {
+            store = getImapConnection();
+        }
+        try {
+            mailFolder = store.getFolder(folder);
+            mailFolder.open(Folder.READ_WRITE);
+            SearchTerm searchTerm = new AndTerm(new SubjectTerm(emailSubject), new BodyTerm(emailSubject));
+            Message[] messages = mailFolder.search(searchTerm);
+            for (Message message : messages) {
+                if (message.getSubject().contains(emailSubject)) {
+                    log.info("Found the Email with Subject : " + emailSubject);
+                    emailReceived = true;
+                    break;
+                }
+            }
+        } finally {
+            if (store != null) {
+                store.close();
+            }
+        }
+        return emailReceived;
+    }
+
+    /**
+     * Check inbox and make sure a particular email is deleted.
+     *
+     * @param emailSubject
+     * @param protocol
+     * @return
+     * @throws ESBMailTransportIntegrationTestException
+     */
+    public static boolean checkEmailDeleted(String emailSubject, String protocol)
+            throws ESBMailTransportIntegrationTestException, MessagingException {
+        boolean isEmailDeleted = false;
+        long startTime = System.currentTimeMillis();
+
+        while ((System.currentTimeMillis() - startTime) < WAIT_TIME_MS) {
+            if (!isMailReceivedBySubject(emailSubject, EMAIL_INBOX, protocol)) {
+                log.info("Email has been deleted successfully!");
+                isEmailDeleted = true;
+                break;
+            }
+            try {
+                Thread.sleep(500);
+            } catch (InterruptedException e) {
+                log.warn("Error during thread sleep for 0.5 sec");
```

Comment:
I don't think we need to have a warning log, since we'll anyway not be breaking the loop. WDYT?

Reply:
But isn't this the easiest way to know if an exception occurs at thread sleep?

Reply:
The correct way would be 

try {
  Thread.sleep(100);
} catch (InterruptedException ex) {
  Thread.currentThread().interrupt(); // Here!
  throw new RuntimeException(ex);
}

Going by
http://www.yegor256.com/2015/10/20/interrupted-exception.html

Reply:
@Thishani, do we need to know if an exception occurs? We just need the loop to continue right? The only impact is that it'll reduce the time taken for that particular 0.5 seconds.

@Hasitha, we certainly do not want to throw a runtime or to interrupt the current thread here. We need the loop to continue.

Reply:
yes, I guess we can ignore the try-catch. Removed it.